src/client.rs

```diff
@@ -912,16 +915,64 @@ impl Client {
         // Imagine if a cookie is set with path=/download/some_identifier. How do we get that
         // cookie without triggering a request for the (large) file? I don't know. Hence: TODO.
         let old_url = self.current_url_().await?;
-        let url = old_url.clone().join(&url)?;
-        let cookie_url = url.clone().join("/please_give_me_your_cookies")?;
-        self.goto(cookie_url.as_str()).await?;
+        let cookie_url = old_url.join(url)?.join("/please_give_me_your_cookies")?;
+        self.with_raw_cookie_client_for(method, url, Some(cookie_url.as_str()), before)
+            .await
+    }
+
+    /// Issue an HTTP request to the given `url` with all the same cookies as the current
+    /// session. If set, the client will first navigate to
+    /// `cookie_url` to retrieve cookies.
+    ///
+    /// Calling this method is equivalent to calling `with_raw_cookie_client_for` with an empty closure.
+    ///
+    /// `raw_client_for` is an alternative that automatically handles cookies.
+    pub async fn raw_cookie_client_for(
+        &self,
+        method: Method,
+        url: &str,
+        cookie_url: Option<&str>,
+    ) -> Result<hyper::Response<hyper::Body>, error::CmdError> {
+        self.with_raw_cookie_client_for(method, url, cookie_url, |req| {
+            req.body(hyper::Body::empty()).unwrap()
+        })
+        .await
+    }
+
+    /// Build and issue an HTTP request to the given `url` with all the same cookies as the current
+    /// session. If set, the client will first navigate to
+    /// `cookie_url` to retrieve cookies.
+    ///
+    /// Before the HTTP request is issued, the given `before` closure will be called with a handle
+    /// to the `Request` about to be sent.
+    ///
+    /// `with_raw_client_for` is an alternative that automatically handles cookies.
+    pub async fn with_raw_cookie_client_for<F>(
+        &self,
+        method: Method,
+        url: &str,
+        cookie_url: Option<&str>,
+        before: F,
+    ) -> Result<hyper::Response<hyper::Body>, error::CmdError>
+    where
+        F: FnOnce(http::request::Builder) -> hyper::Request<hyper::Body>,
+    {
+        let old_url = self.current_url_().await?;
+        let url = old_url.clone().join(url)?;
+
+        if let Some(cookie_url) = cookie_url {
+            self.goto(cookie_url).await?;
+            self.back().await?;
```

Comment:
Ah, this doesn't work. We need to gather the cookie _before_ we go back. That's the whole reason why the cookie URL is needed in the first place. `GetCookies` _only_ gives you the cookies for whatever the _current_ URL is, which is why we need to navigate to the target domain before calling it.

Reply:
Ah yeah, good point. Will fix.